src/tools/rbd_mirror/InstanceWatcher.h

```diff
@@ -4,54 +4,74 @@
 #ifndef CEPH_RBD_MIRROR_INSTANCE_WATCHER_H
 #define CEPH_RBD_MIRROR_INSTANCE_WATCHER_H
 
+#include <map>
+#include <set>
 #include <string>
 #include <vector>
 #include <boost/optional.hpp>
 
+#include "common/AsyncOpTracker.h"
 #include "librbd/Watcher.h"
 #include "librbd/managed_lock/Types.h"
+#include "tools/rbd_mirror/instance_watcher/Types.h"
 
 namespace librbd {
-  class ImageCtx;
-  template <typename> class ManagedLock;
+
+class ImageCtx;
+template <typename> class ManagedLock;
+
 }
 
 namespace rbd {
 namespace mirror {
 
+template <typename> class InstanceReplayer;
+template <typename> struct Threads;
+
 template <typename ImageCtxT = librbd::ImageCtx>
 class InstanceWatcher : protected librbd::Watcher {
 public:
   static void get_instances(librados::IoCtx &io_ctx,
                             std::vector<std::string> *instance_ids,
                             Context *on_finish);
-  static void remove_instance(librados::IoCtx &io_ctx, ContextWQ *work_queue,
+  static void remove_instance(Threads<ImageCtxT> *threads,
+                              librados::IoCtx &io_ctx,
                               const std::string &instance_id,
                               Context *on_finish);
 
   static InstanceWatcher *create(
-    librados::IoCtx &io_ctx, ContextWQ *work_queue,
-    const boost::optional<std::string> &id = boost::none) {
-    return new InstanceWatcher(io_ctx, work_queue, id);
+    Threads<ImageCtxT> *threads, librados::IoCtx &io_ctx,
+    InstanceReplayer<ImageCtxT> *instance_replayer) {
+    return new InstanceWatcher(threads, io_ctx, instance_replayer, boost::none);
   }
   void destroy() {
     delete this;
   }
 
-  InstanceWatcher(librados::IoCtx &io_ctx, ContextWQ *work_queue,
-                  const boost::optional<std::string> &id = boost::none);
+  InstanceWatcher(Threads<ImageCtxT> *threads, librados::IoCtx &io_ctx,
+                  InstanceReplayer<ImageCtxT> *instance_replayer,
+                  const boost::optional<std::string> &id);
```

Comment:
Do you mean to remove `id` argument, or to make it non-optional?

The `id` is mostly needed only for `InstanceWatcher::remove_instance()` (and may be for tests). So it is not for public use (and `create` method does not have it).

I think I can make it non-optional in the constructor and properly set in `create()`.